File: src/main/java/io/github/svegon/capi/event/network/ServerConnectCallback.java
package io.github.svegon.capi.event.network;

import net.fabricmc.fabric.api.event.Event;
import net.fabricmc.fabric.api.event.EventFactory;
import net.minecraft.client.MinecraftClient;
import net.minecraft.client.gui.screen.Screen;
import net.minecraft.client.network.ServerAddress;
import net.minecraft.client.network.ServerInfo;
import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;

@FunctionalInterface
public interface ServerConnectCallback {
    Event<ServerConnectCallback> EVENT = EventFactory.createArrayBacked(ServerConnectCallback.class,
            (screen, client, address, info, quickPlay, ci) -> {},
            listeners -> (screen, client, address, info, quickPlay, ci) -> {
        for (ServerConnectCallback listener : listeners) {
            listener.onServerConnect(screen, client, address, info, quickPlay, ci);

            if (ci.isCancelled()) {
                return;
            }
        }
    });

    void onServerConnect(Screen screen, MinecraftClient client, ServerAddress address, ServerInfo info,
                         boolean quickPlay, CallbackInfo ci);
}

File: src/main/kotlin/io/github/svegon/mclientapi/event/network/ServerConnectCallback.kt
package io.github.svegon.capi.event.network

import net.fabricmc.fabric.api.event.Event
import net.fabricmc.fabric.api.event.EventFactory
import net.minecraft.client.MinecraftClient

fun interface ServerConnectCallback {
    fun onServerConnect(
        screen: Screen?, client: MinecraftClient?, address: ServerAddress?, info: ServerInfo?,
        quickPlay: Boolean, ci: CallbackInfo?
    )

    companion object {
        @JvmField
        val EVENT: Event<ServerConnectCallback> = EventFactory.createArrayBacked<ServerConnectCallback>(
            ServerConnectCallback::class.java,
            ServerConnectCallback { screen: Screen?, client: MinecraftClient?, address: ServerAddress?, info: ServerInfo?, quickPlay: Boolean, ci: CallbackInfo? -> }
        ) { listeners: Array<ServerConnectCallback> ->
            ServerConnectCallback { screen: Screen?, client: MinecraftClient?, address: ServerAddress?, info: ServerInfo?, quickPlay: Boolean, ci: CallbackInfo ->
                for (listener in listeners) {
                    listener.onServerConnect(screen, client, address, info, quickPlay, ci)

                    if (ci.isCancelled()) {
                        return@ServerConnectCallback
                    }
                }
            }
        }
    }
}
